pkg/webhook/webhook.go

```diff
@@ -0,0 +1,219 @@
+package webhook
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"regexp"
+
+	gogsclient "github.com/gogits/go-gogs-client"
+	"github.com/gorilla/mux"
+	v1controller "github.com/rancher/gitjob/pkg/generated/controllers/gitjob.cattle.io/v1"
+	"github.com/rancher/gitjob/pkg/types"
+	"github.com/rancher/steve/pkg/aggregation"
+	corev1controller "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/go-playground/webhooks.v5/bitbucket"
+	bitbucketserver "gopkg.in/go-playground/webhooks.v5/bitbucket-server"
+	"gopkg.in/go-playground/webhooks.v5/github"
+	"gopkg.in/go-playground/webhooks.v5/gitlab"
+	"gopkg.in/go-playground/webhooks.v5/gogs"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	webhookSecretName = "gitjob-webhook"
+
+	githubKey          = "github"
+	gitlabKey          = "gitlab"
+	bitbucketKey       = "bitbucket"
+	bitbucketServerKey = "bitbucket-server"
+	gogsKey            = "gogs"
+)
+
+type Webhook struct {
+	gitjobs   v1controller.GitJobController
+	secrets   corev1controller.SecretController
+	namespace string
+
+	github          *github.Webhook
+	gitlab          *gitlab.Webhook
+	bitbucket       *bitbucket.Webhook
+	bitbucketServer *bitbucketserver.Webhook
+	gogs            *gogs.Webhook
+}
+
+func New(ctx context.Context, rContext *types.Context) *Webhook {
+	webhook := &Webhook{
+		gitjobs:   rContext.Gitjob.Gitjob().V1().GitJob(),
+		secrets:   rContext.Core.Core().V1().Secret(),
+		namespace: rContext.Namespace,
+	}
+
+	rContext.Core.Core().V1().Secret().OnChange(ctx, "webhook-secret", webhook.onSecretChange)
+	return webhook
+}
+
+func (w *Webhook) onSecretChange(key string, secret *corev1.Secret) (*corev1.Secret, error) {
+	if secret == nil || secret.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if secret.Name != webhookSecretName && secret.Namespace != w.namespace {
+		return nil, nil
+	}
+
+	var err error
+	w.github, err = github.New(github.Options.Secret(string(secret.Data[githubKey])))
```

Comment:
What happens when a user does not have one of these keys?

Reply:
Good question. I checked all the function and it doesn't return error.
```
func (WebhookOptions) Secret(secret string) Option {
	return func(hook *Webhook) error {
		hook.secret = secret
		return nil
	}
}
```

If user doesn't have the key, we just don't validate the payload. This means you have no protection against payload from world. Also see https://argoproj.github.io/argo-cd/operator-manual/webhook/

Reply:
Ah I see. I like the implementation. It was not obvious at first, but that makes sense.